File: client/pkg/capsule/runplan.go
package capsule

import (
	pb "github.com/onescluster/coordinator/pkg/proto"
)

// RunPlan is the normalized execution plan shared between runtimes.
// Mirrors libadep-core's runplan.rs shape.
type RunPlan struct {
	CapsuleID       string           `json:"capsule_id"`
	Name            string           `json:"name"`
	Version         string           `json:"version"`
	Docker          *DockerRuntime   `json:"docker,omitempty"`
	Native          *NativeRuntime   `json:"native,omitempty"`
	PythonUv        *PythonUvRuntime `json:"python_uv,omitempty"`
	CPUCores        *uint32          `json:"cpu_cores,omitempty"`
	MemoryBytes     *uint64          `json:"memory_bytes,omitempty"`
	GPUProfile      string           `json:"gpu_profile,omitempty"`
	EgressAllowlist []string         `json:"egress_allowlist,omitempty"`
}

type DockerRuntime struct {
	Image      string            `json:"image"`
	Digest     string            `json:"digest,omitempty"`
	Command    []string          `json:"command,omitempty"`
	Env        map[string]string `json:"env,omitempty"`
	WorkingDir string            `json:"working_dir,omitempty"`
	User       string            `json:"user,omitempty"`
	Ports      []Port            `json:"ports,omitempty"`
	Mounts     []Mount           `json:"mounts,omitempty"`
}

type NativeRuntime struct {
	BinaryPath string            `json:"binary_path"`
	Args       []string          `json:"args,omitempty"`
	Env        map[string]string `json:"env,omitempty"`
	WorkingDir string            `json:"working_dir,omitempty"`
}

type PythonUvRuntime struct {
	Entrypoint string            `json:"entrypoint"`
	Args       []string          `json:"args,omitempty"`
	Env        map[string]string `json:"env,omitempty"`
	WorkingDir string            `json:"working_dir,omitempty"`
	Ports      []Port            `json:"ports,omitempty"`
}

type Port struct {
	ContainerPort uint32  `json:"container_port"`
	HostPort      *uint32 `json:"host_port,omitempty"`
	Protocol      string  `json:"protocol,omitempty"`
}

type Mount struct {
	Source   string `json:"source"`
	Target   string `json:"target"`
	Readonly bool   `json:"readonly"`
}

// ToRunPlan converts a CapsuleManifest into a RunPlan (v0).
func (m *CapsuleManifest) ToRunPlan() (*RunPlan, error) {
	// Validate basic fields are present
	if m.Name == "" {
		return nil, ErrInvalidName
	}
	if m.Version == "" {
		return nil, ErrInvalidVersion
	}

	rp := &RunPlan{
		CapsuleID: m.Name,
		Name:      m.Name,
		Version:   m.Version,
	}

	// Memory hint from vram_min if present
	if m.Requirements.VRAMMin != "" {
		if bytes, err := parseMemoryString(m.Requirements.VRAMMin); err == nil {
			b := uint64(bytes)
			rp.MemoryBytes = &b
		}
	}

	// Ports helper
	var ports []Port
	if m.Execution.Port > 0 {
		ports = append(ports, Port{ContainerPort: uint32(m.Execution.Port), Protocol: "tcp"})
	}

	switch m.Execution.Runtime {
	case RuntimeDocker:
		rp.Docker = &DockerRuntime{
			Image:   m.Execution.Entrypoint,
			Command: []string{},
			Env:     cloneEnv(m.Execution.Env),
			Ports:   ports,
			Mounts:  []Mount{},
		}
	case RuntimeNative:
		rp.Native = &NativeRuntime{
			BinaryPath: m.Execution.Entrypoint,
			Args:       []string{},
			Env:        cloneEnv(m.Execution.Env),
		}
	case RuntimePythonUv:
		rp.PythonUv = &PythonUvRuntime{
			Entrypoint: m.Execution.Entrypoint,
			Args:       []string{},
			Env:        cloneEnv(m.Execution.Env),
			Ports:      ports,
		}
	default:
		return nil, ErrInvalidRuntime
	}

	return rp, nil
}

func cloneEnv(env map[string]string) map[string]string {
	if env == nil {
		return nil
	}
	out := make(map[string]string, len(env))
	for k, v := range env {
		out[k] = v
	}
	return out
}

// ToProto converts internal RunPlan to proto RunPlan (coordinator.proto)
func (rp *RunPlan) ToProto() *pb.RunPlan {
	if rp == nil {
		return nil
	}

	proto := &pb.RunPlan{
		CapsuleId:       rp.CapsuleID,
		Name:            rp.Name,
		Version:         rp.Version,
		GpuProfile:      rp.GPUProfile,
		EgressAllowlist: rp.EgressAllowlist,
	}

	if rp.CPUCores != nil {
		proto.CpuCores = *rp.CPUCores
	}
	if rp.MemoryBytes != nil {
		proto.MemoryBytes = *rp.MemoryBytes
	}

	if rp.Docker != nil {
		ports := make([]*pb.Port, len(rp.Docker.Ports))
		for i, p := range rp.Docker.Ports {
			ports[i] = &pb.Port{
				ContainerPort: p.ContainerPort,
				Protocol:      p.Protocol,
			}
			if p.HostPort != nil {
				ports[i].HostPort = *p.HostPort
			}
		}

		mounts := make([]*pb.Mount, len(rp.Docker.Mounts))
		for i, m := range rp.Docker.Mounts {
			mounts[i] = &pb.Mount{
				Source:   m.Source,
				Target:   m.Target,
				Readonly: m.Readonly,
			}
		}

		proto.Runtime = &pb.RunPlan_Docker{
			Docker: &pb.DockerRuntime{
				Image:      rp.Docker.Image,
				Digest:     rp.Docker.Digest,
				Command:    rp.Docker.Command,
				Env:        rp.Docker.Env,
				WorkingDir: rp.Docker.WorkingDir,
				User:       rp.Docker.User,
				Ports:      ports,
				Mounts:     mounts,
			},
		}
	} else if rp.Native != nil {
		proto.Runtime = &pb.RunPlan_Native{
			Native: &pb.NativeRuntime{
				BinaryPath: rp.Native.BinaryPath,
				Args:       rp.Native.Args,
				Env:        rp.Native.Env,
				WorkingDir: rp.Native.WorkingDir,
			},
		}
	} else if rp.PythonUv != nil {
		ports := make([]*pb.Port, len(rp.PythonUv.Ports))
		for i, p := range rp.PythonUv.Ports {
			ports[i] = &pb.Port{
				ContainerPort: p.ContainerPort,
				Protocol:      p.Protocol,
			}
			if p.HostPort != nil {
				ports[i].HostPort = *p.HostPort
			}
		}

		proto.Runtime = &pb.RunPlan_PythonUv{
			PythonUv: &pb.PythonUvRuntime{
				Entrypoint: rp.PythonUv.Entrypoint,
				Args:       rp.PythonUv.Args,
				Env:        rp.PythonUv.Env,
				WorkingDir: rp.PythonUv.WorkingDir,
				Ports:      ports,
			},
		}
	}

	return proto
}

File: libadep/core/src/runplan.rs
use alloc::collections::BTreeMap;
use alloc::string::String;
use alloc::vec::Vec;
use serde::{Deserialize, Serialize};
use std::collections::HashMap;

use crate::capsule_v1::{CapsuleManifestV1, RuntimeType};
use crate::error::CapsuleError;

const DEFAULT_STORAGE_MOUNT_BASE: &str = "/var/lib/gumball/volumes";

fn default_storage_mount_base() -> String {
    std::env::var("GUMBALL_STORAGE_BASE").unwrap_or_else(|_| DEFAULT_STORAGE_MOUNT_BASE.to_string())
}

/// Normalized execution plan produced from capsule_v1 manifests.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct RunPlan {
    pub capsule_id: String,
    pub name: String,
    pub version: String,

    #[serde(flatten)]
    pub runtime: RunPlanRuntime,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub cpu_cores: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub memory_bytes: Option<u64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub gpu_profile: Option<String>,

    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub egress_allowlist: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum RunPlanRuntime {
    #[serde(rename = "docker")]
    Docker(DockerRuntime),
    #[serde(rename = "native")]
    Native(NativeRuntime),
    #[serde(rename = "python_uv")]
    PythonUv(PythonUvRuntime),
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct DockerRuntime {
    pub image: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub digest: Option<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub command: Vec<String>,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub env: BTreeMap<String, String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub working_dir: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub user: Option<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub ports: Vec<Port>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub mounts: Vec<Mount>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct NativeRuntime {
    pub binary_path: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub args: Vec<String>,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub env: BTreeMap<String, String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub working_dir: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct PythonUvRuntime {
    pub entrypoint: String,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub args: Vec<String>,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub env: BTreeMap<String, String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub working_dir: Option<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub ports: Vec<Port>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct Port {
    pub container_port: u32,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub host_port: Option<u32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub protocol: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
pub struct Mount {
    pub source: String,
    pub target: String,
    pub readonly: bool,
}

impl CapsuleManifestV1 {
    /// Convert a validated capsule_v1 manifest into a normalized RunPlan.
    pub fn to_run_plan(&self) -> Result<RunPlan, CapsuleError> {
        let ports = port_list(self.execution.port);
        let env = ordered_env(&self.execution.env);

        let runtime = match self.execution.runtime {
            RuntimeType::Docker => {
                let mut mounts = Vec::new();
                if !self.storage.volumes.is_empty() {
                    let base = default_storage_mount_base();
                    for vol in &self.storage.volumes {
                        let name = vol.name.trim();
                        let mount_path = vol.mount_path.trim();
                        if name.is_empty() || mount_path.is_empty() || !mount_path.starts_with('/') || mount_path.contains("..") {
                            return Err(CapsuleError::ValidationError(
                                "invalid storage volume (requires name and absolute mount_path)".to_string(),
                            ));
                        }

                        mounts.push(Mount {
                            source: format!("{}/{}/{}", base.trim_end_matches('/'), self.name, name),
                            target: mount_path.to_string(),
                            readonly: vol.read_only,
                        });
                    }
                }

                RunPlanRuntime::Docker(DockerRuntime {
                image: self.execution.entrypoint.clone(),
                digest: None,
                command: Vec::new(),
                env: env.clone(),
                working_dir: None,
                user: None,
                ports: ports.clone(),
                mounts,
            })
            }
            RuntimeType::Native => RunPlanRuntime::Native(NativeRuntime {
                binary_path: self.execution.entrypoint.clone(),
                args: Vec::new(),
                env: env.clone(),
                working_dir: None,
            }),
            RuntimeType::PythonUv => RunPlanRuntime::PythonUv(PythonUvRuntime {
                entrypoint: self.execution.entrypoint.clone(),
                args: Vec::new(),
                env: env.clone(),
                working_dir: None,
                ports: ports.clone(),
            }),
        };

        // storage validation is handled by CapsuleManifestV1::validate(); keep to_run_plan focused.

        let memory_bytes = self.requirements.vram_min_bytes()?;

        Ok(RunPlan {
            capsule_id: self.name.clone(),
            name: self.name.clone(),
            version: self.version.clone(),
            runtime,
            cpu_cores: None,
            memory_bytes,
            gpu_profile: None,
            egress_allowlist: Vec::new(),
        })
    }
}

fn ordered_env(env: &HashMap<String, String>) -> BTreeMap<String, String> {
    env.iter().map(|(k, v)| (k.clone(), v.clone())).collect()
}

fn port_list(port: Option<u16>) -> Vec<Port> {
    port
        .map(|p| Port {
            container_port: p as u32,
            host_port: None,
            protocol: Some("tcp".to_string()),
        })
        .into_iter()
        .collect()
}

#[cfg(test)]
mod tests {
    use super::*;

    const SAMPLE_PYTHON_TOML: &str = r#"
schema_version = "1.0"
name = "mlx-qwen3-8b"
version = "1.0.0"
type = "inference"

[execution]
runtime = "python-uv"
entrypoint = "server.py"
port = 8081

[execution.env]
GUMBALL_MODEL = "qwen3-8b"

[capabilities]
chat = true
function_calling = true
vision = false
context_length = 8192

[model]
source = "hf:org/model"
"#;

    const SAMPLE_DOCKER_TOML: &str = r#"
schema_version = "1.0"
name = "hello-docker"
version = "0.1.0"
type = "app"

[execution]
runtime = "docker"
entrypoint = "ghcr.io/example/hello:latest"
port = 8080
"#;

    #[test]
    fn runplan_from_python_uv_manifest() {
        let manifest = CapsuleManifestV1::from_toml(SAMPLE_PYTHON_TOML).unwrap();
        manifest.validate().unwrap();
        let plan = manifest.to_run_plan().unwrap();

        let json = serde_json::to_value(&plan).unwrap();
        let expected = serde_json::json!({
            "capsule_id": "mlx-qwen3-8b",
            "name": "mlx-qwen3-8b",
            "version": "1.0.0",
            "python_uv": {
                "entrypoint": "server.py",
                "env": {"GUMBALL_MODEL": "qwen3-8b"},
                "ports": [
                    {"container_port": 8081, "protocol": "tcp"}
                ]
            }
        });

        assert_eq!(json, expected);
    }

    #[test]
    fn runplan_from_docker_manifest() {
        let manifest = CapsuleManifestV1::from_toml(SAMPLE_DOCKER_TOML).unwrap();
        manifest.validate().unwrap();
        let plan = manifest.to_run_plan().unwrap();

        let json = serde_json::to_value(&plan).unwrap();
        let expected = serde_json::json!({
            "capsule_id": "hello-docker",
            "name": "hello-docker",
            "version": "0.1.0",
            "docker": {
                "image": "ghcr.io/example/hello:latest",
                "ports": [
                    {"container_port": 8080, "protocol": "tcp"}
                ]
            }
        });

        assert_eq!(json, expected);
    }
}
